Multicursor reformatting (2)

pbwolf opened the pull request:
## What has changed?

This branch experimentally restores multi-cursor reformatting (based on the branch of PR #2741) with additional enhancements to formatPosition's cursor repositioning. 

Issue 2748 Case 1 (The cursor jumped to the next line when it had been in whitespace that the reformatter revised) is resolved by making a distinct whitespace edit for each line instead of combining spaces and newlines into a single edit.

Issue 2748 Case 2 (The cursor, at the beginning of a blank line, did not move to the indentation tab-stop): 
- Case 2a, when the formatter should insert the necessary spaces, is resolved by passing an actual cursor position as the reference point to the formatter. If there is more than one cursor then one is chosen arbitrarily. The non-arbitrary technique of using the beginning of the form as the reference point did not recognize that the formatter might trim (not indent) blank lines other than the one with the reference point.
- Case 2b, when the document already had enough indentation spaces but they were to the right of the cursor, is mitigated, for the single-cursor case only, by asynchronously imposing the formatter's newIndex as the selection if it is different from where the cursor floated to naturally.

Reformatting between two top-level forms, or performing a structural edit that changed a top-level form's delimiters, reformatted the whole document using an interface that did not specify a reference point. That occasioned a few other adjustments on this branch:
- This branch resorts to whole-document formatting less often, by narrowing the post-structural-edit reformat radius to the form (not its enclosing list) if there is no enclosing list. Also, this branch formats a single-form document as a form, not a document.
- This branch formats a document even if its first line was a comment.

I have tried this branch only in Linux.

Fixes #2748

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Made sure there is an issue registered with a clear problem statement that this PR addresses, (created the issue if it was not present).
    - [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [ ] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [x] Tests
  - [x] Tested the particular change
  - [ ] Figured if the change might have some side effects and tested those as well.
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

Ping @pez, @bpringe, @corasaurus-hex, @Cyrik

pbwolf committed:
Subdivide reformatting whitespace edits at newline

pbwolf committed:
formatDocIndexRange using [-1,-1] as whole-doc sentinel to distinguish from a single form that happens to be coterminous with the document

pbwolf committed:
Reapply "Merge pull request #2741 from pbwolf/2738-multicursor-reformat-C"

This reverts commit 458d9d4676697fe702923d4623166baee8fb477c.

pbwolf committed:
Merge branch '2748-reformat-positioning' into 2748-reformat-positioning.2

pbwolf committed:
Auto-reformat: if no form encloses the offset of interest, fall back to the form of that offset

pbwolf committed:
Waive rule about not reindenting the top level form at the beginning of the range to reformat, in the case where the range is the whole document

pbwolf committed:
Move comment closer to applicable code

pbwolf committed:
Prettier

pbwolf committed:
Impose the formatter's newIndex cursor position, if there is only one cursor

pbwolf committed:
Unit test: reformatting edits keep the cursor on the same line when adjusting whitespace

pbwolf committed:
Merge branch 'dev' into 2748-reformat-positioning.2

pbwolf committed:
Entry fixing 2748

pbwolf commented:
Merged developments from dev, added an entry to CHANGELOG, and edited the description to mention that I have tried it in Linux only

PEZ commented:
Awesome. I am testing and can confirm that the regressions are gone. And again that I love multi-cursor formatting!

There's another thing though. I'm noting that we're not placing cursors 2 - n correctly after formatting. Here's an example:

```clojure
{:a :1|  ; cursor 1
 :b :2|  ; cursor 2
 :c :3|  ; cursor 3
 :d :4|} ; cursor 4
```

<kbd>enter</kbd>

```clojure
{:a :1
|
 :b :2
 |
 :c :3
 |
 :d :4
 |}
```

This is as before, because some bug in the indenter, but then we have:


<kbd>tab</kbd>

```clojure
{:a :1
 |
 :b :2
|
 :c :3
|
 :d :4|}
```

Which we couldn't really do before. I'd say that if this is not trivial to fix, we can still merge, and fix later. But if you think it is easy, then better fix before merge. WDYS?

pbwolf committed:
Format Current Form hereby "indents" blank lines where a cursor is located, by
(a) inserting missing spaces and
(b) shuttling the cursor to the indentation stop even if the necessary spaces were already in the document.
TODO: find a new home in the integration test for obsoleted tests of new-index that were in formatter_test.cljs.

pbwolf commented:
formatter.cljs is revised to facilitate indentation and cursor positioning on blank lines where any cursor is located. Consequently, calva-fmt no longer computes a new cursor position. (VS Code automatically shuttles cursors after edits, such as the formatter's, more smoothly than Calva can do asynchronously after the edits commit.) Therefore, unit tests of "new-index" under various interesting circumstances are removed. They might find a new home in an integration test. Unit tests of indentation of text are revised to use text-based heuristics (direct observation) instead of relying on "new-index". 

The crucial, clever kluge in formatter.cljs remains unchanged: it tricks clj-fmt into indenting blank-lines-of-interest by putting a temporary token at the end of those lines. This can apparently lead clj-fmt to over-indent lines-of-code that follow an overly-indented blank line where a cursor was located. This irregularity may be the price of not disturbing multi-line strings, where space is significant.

pbwolf committed:
Merged dev

pbwolf committed:
Stake wider reformat range, zap blan*k-line kluge

formatPosition can now pass all cursors plus a range-defining index
(e.g., the location of a form's opening parenthesis)
to the formatter, so it no longer needs to pass a single cursor
in place of the range-defining index.

pbwolf committed:
Merge dev

pbwolf committed:
Remove disused code

pbwolf committed:
No changes: spark a re-run of integration tests

pbwolf committed:
Merge branch 'dev' into 2748-reformat-positioning.2

pbwolf committed:
Merge branch 'dev' into 2748-reformat-positioning.2

PEZ commented on CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@ Changes to Calva.
 
 ## [Unreleased]
 
+- Fix: [Reformat all edited locations after multicursor structural edit](https://github.com/BetterThanTomorrow/calva/issues/2738), [Wrong cursor position when using **Calva Format: Format Current Form**](https://github.com/BetterThanTomorrow/calva/issues/2748)
```
```suggestion
- Fix: [Reformat all edited locations after multicursor structural edit](https://github.com/BetterThanTomorrow/calva/issues/2738)
- Fix: [Wrong cursor position when using **Calva Format: Format Current Form**](https://github.com/BetterThanTomorrow/calva/issues/2748)
```

PEZ committed:
Update CHANGELOG.md

PEZ commented:
Thanks, @pbwolf ! 🙏 Multi-cursor formatting rocks!

PEZ commented:
Now I have tried this on Windows, and died inside a bit. Not because the formatting didn't work. It works beautifully. Will release now! 🎉
